telemetry/tracks/class-tracks-client.php

```diff
@@ -0,0 +1,187 @@
+	 *
+	 * @var ?Tracks_Client
+	 */
+	protected static $instance = null;
+
+	/**
+	 * Events queued to be sent to the Tracks pixel.
+	 *
+	 * @var array<Tracks_Event>
+	 */
+	protected $events = array();
+
+	/**
+	 * Constructor.
+	 */
+	private function __construct() {
+		// Track events asynchronously (inject pixels in the footer).
+		add_action( 'admin_footer', array( $this, 'render_tracking_pixels' ) );
+		// Synchronously track any remaining events that could not be added to
+		// the footer.
+		add_action( 'shutdown', array( $this, 'record_remaining_events' ) );
+	}
+
+	/**
+	 * Instantiates the singleton.
+	 *
+	 * @return Tracks_Client
+	 */
+	public static function instance(): Tracks_Client {
+		if ( is_null( static::$instance ) ) {
+			static::$instance = new Tracks_Client();
+		}
+
+		return static::$instance;
+	}
+
+	/**
+	 * Enqueues an event to be recorded asynchronously.
+	 *
+	 * @param Tracks_Event $event The event to record.
+	 * @return bool|WP_Error True if the event was enqueued for recording.
+	 *                       False if the event is not recordable.
+	 *                       WP_Error if the event is generating an error.
+	 */
+	public function record_event_asynchronously( Tracks_Event $event ) {
+		$is_event_recordable = $event->is_recordable();
+
+		if ( true !== $is_event_recordable ) {
+			return $is_event_recordable;
+		}
+
+		static::instance()->events[] = $event;
+
+		return true;
+	}
+
+	/**
+	 * Records an event synchronously (using a GET request).
+	 *
+	 * @param Tracks_Event $event The event to record.
+	 * @return bool|WP_Error True if recording the event succeeded.
+	 *                       False if the event is not recordable.
+	 *                       WP_Error if any error occurred.
+	 */
+	public function record_event_synchronously( Tracks_Event $event ) {
+		$is_event_recordable = $event->is_recordable();
+
+		if ( true !== $is_event_recordable ) {
+			return $is_event_recordable;
+		}
+
+		$pixel_url = static::instance()->generate_pixel_url( $event );
+
+		if ( null === $pixel_url ) {
+			return new WP_Error(
+				'invalid_pixel',
+				'cannot generate tracks pixel for given input',
+				array( 'status' => 400 )
+			);
+		}
+
+		// Add the Request Timestamp and URL terminator just before the HTTP
+		// request.
+		$pixel_url .= '&_rt=' . static::build_timestamp() . '&_=_';
+
+		$request = wp_safe_remote_get(
+			$pixel_url,
+			array(
+				'blocking'    => false,
+				'redirection' => 2,
+				'httpversion' => '1.1',
+				'timeout'     => 1,
+			)
+		);
+
+		if ( is_wp_error( $request ) ) {
+			return $request;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Outputs a Tracks pixel for every registered event.
+	 */
+	public function render_tracking_pixels(): void {
+		foreach ( $this->events as $event ) {
+			$pixel_url = $this->generate_pixel_url( $event );
+
+			if ( null === $pixel_url ) {
+				continue;
+			}
+
+			echo '<img style="position: fixed;" src="', esc_url( $pixel_url ), '" />';
```

Comment:
Is there a benefit of doing this vs simply running on shutdown? Suppose we go overboard and track 20 different things during the same page load.

Reply:
Agreed on this, I'm not fully grasping why this would be better over running on shutdown 🤔. We want the site's experience to not be impacted, even if that may come at the expense of telemetry imo.

Reply:
1. A benefit for having this on top of the shutdown part is that any events that are recorded before the pixel are rendered will be sent first so if we have 20 different events to track, there is a change those events could be sent in 2 batches. One for events before the pixel is rendered and one for events after.
2. Agreed it is kind of a misnomer but the naming convention is consistent between the Parsely plugin, WPCOM and Woo

Reply:
I'm leaning towards starting with just doing this on shutdown, rather than the pixel approach and shutdown approach. I think we should try to minimize our impact on the page as much as possible. Considering this is telemetry, having it go on shutdown at the very should be fine imo.